sdk/python/feast/client.py

```diff
@@ -508,36 +527,125 @@ def ingest(
+        time_start = time.time()
 
-        # Update the feature set based on DataFrame schema
-        if force_update:
-            # Use a small as reference DataFrame to infer fields
-            ref_df = table.to_batches(max_chunksize=20)[0].to_pandas()
+        # Read table and get row count
+        tmp_table_name = _read_table_from_source(source, chunk_size,
+                                                 max_workers)
+
+        pq_file = pq.ParquetFile(tmp_table_name)
+
+        # Delete source from memory after conversion into PyArrow table
+        del source
+
+        row_count = pq_file.metadata.num_rows
 
-            feature_set.infer_fields_from_df(
-                ref_df, discard_unused_fields=True, replace_existing_features=True
+        # Update the feature set based on PyArrow table of the first row group
+        if force_update:
+            feature_set.infer_fields_from_pa(
+                table=pq_file.read_row_group(0),
+                discard_unused_fields=True,
+                replace_existing_features=True
             )
             self.apply(feature_set)
 
         feature_set = self.get_feature_set(name, version)
 
+        time_end = time.time()
+        exec_time = round(time_end - time_start, 2)
+        print(f"Took {exec_time}s to read, validate and split parquet file.")
+
+        # Kafka configs
+        brokers = feature_set.get_kafka_source_brokers()
+        topic = feature_set.get_kafka_source_topic()
+        producer = get_producer(brokers)
+
+        # Progress bar will always display average rate
+        pbar = tqdm(total=row_count,
+                    unit="rows",
+                    smoothing=0,
+                    disable=disable_progress_bar)
+
+        # Loop optimization declarations
+        produce = producer.produce
+        flush = producer.flush
+        update = pbar.update
+
+        # Define tracker statistic context
+        ctx = {"success_count": 0, "error_count": 0, "last_exception": ""}
+
+        # Transform and push data to Kafka
         if feature_set.source.source_type == "Kafka":
-            ingest_table_to_kafka(
-                feature_set=feature_set,
-                table=table,
-                max_workers=max_workers,
-                disable_pbar=disable_progress_bar,
-                chunk_size=chunk_size,
-                timeout=timeout,
-            )
+            for chunk in get_feature_row_chunks(
+                    file=tmp_table_name,
+                    row_groups=list(range(pq_file.num_row_groups)),
+                    fs=feature_set,
+                    max_workers=max_workers):
+
+                # Push FeatureRow one chunk at a time to kafka
+                for serialized_row in chunk:
+                    try:
+                        produce(topic=topic, value=serialized_row)
+
+                        # Errors will prevent this progress bar from updating
+                        update(1)
```

Comment:
This update is not connected to a successful push, since it can still fail over the network. Can we connect this back to a callback?

Reply:
The implementation/use of callbacks for `confluent_kafka` and `kafka_python` are different. 

I can remove the use of dictionary as a context to track errors and  shift the context into the `ProducerInterface` class as attributes to be updated.

This shouldn't affect performance too much.

Reply:
But shouldnt we wait for the ack then? You are just incrementing here without waiting for an ack

Reply:
I did a little digging around in the docs and examples, `confluent-kafka` seems to have a per message callback that allows us to check if the message has been successfully delivered or permanently failed (after retries) in their examples.

Callbacks + try-except should be sufficient to catch all exceptions and delivery errors.

Source: https://github.com/confluentinc/confluent-kafka-python/blob/master/examples/producer.py

As for `kafka-python` the default produce `waits for the leader to write the record to its local log only`.  As such, I am not sure if callbacks are sufficient to catch all delivery failures and exceptions.  

The default `retries` parameter is also set to 0, so I am not really sure how un-ACKed or failed message deliveries are handled here.

Source: https://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html

Nonetheless, I've replicated and reuse the original implementation of `kafka-python`'s `send` method in `abstract_producer.py` in the latest commit, do review it and let me know what you think.